packages/core/test/schema.test.ts

```diff
@@ -96,4 +96,22 @@ describe('schema', () => {
 
     expect(m.schema).toEqual(schema);
   });
+
+  it('should infer action types', () => {
+    createMachine({
+      schema: {
+        actions: createSchema<
+          { type: 'greet'; message: string } | { type: 'sayHello' }
+        >()
+      },
+      entry: [
+        { type: 'greet', message: 'hello' },
+        { type: 'sayHello' },
+        // @ts-expect-error
+        { type: 'greet' },
+        // @ts-expect-error
+        { type: 'fake' }
+      ]
```

Comment:
@Andarist This test is failing because the machine is inferring the actions on `entry` here to be part of `TActions`, instead of _only_ inferring `schema.actions` (do we need `NoInfer` here?)

Reply:
> do we need NoInfer here?

This is the only solution to this problem that I know of - so yes. I'm experimenting locally to see how to apply this technique in this situation

Reply:
I've applied this technique here:
https://github.com/statelyai/xstate/commit/b42b0a4cc81a40d253d9aebff43b31a9978d11b6

And it works... sort of. If we only make a single mistake the error is reported at correct position (it doesnt matter which one of those 2 invalid ones I include in the config):
<img width="493" alt="Screenshot 2021-08-20 at 10 47 40" src="https://user-images.githubusercontent.com/9800850/130207158-e013bb55-f0dc-4b43-8126-66a5f7787168.png">
Note that red squiggles are shown because TS reports "Unused '@ts-expect-error' directive, ts(2578)" there as that other action is commented out.

If we include both mistakes the error "bubbles up":
<img width="492" alt="Screenshot 2021-08-20 at 10 49 49" src="https://user-images.githubusercontent.com/9800850/130207682-da03a89f-403d-4b29-8285-211cade335c1.png">
Which is somewhat correct in a sense that this won't allow invalid code but it's not the most user-friendly thing in the world.

The problem is also compounded by the reported error:
<img width="512" alt="Screenshot 2021-08-20 at 10 53 26" src="https://user-images.githubusercontent.com/9800850/130207917-5f8f9be3-044a-40d5-bb01-fde59ab70710.png">

This most likely happens because the message is only reported against the last member of the union, like here ([TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8AoKUA+UDeVSQFxQCIBzAJwgmAIG4oBbCAZwYEMiJ8HgSBLAOyKgBfZGkzZw7QixAAJCABt5AewJCR6LDkmce-GgyX18BAG6sArhGpQwJJWGMQSoVYKqJEAYyW9OUZvhw8GJaxqTklELuQA)):
```ts
type A =
  | { type: "greet"; message: string }
  | { type: "sayHello" }
  | { type: string; some: "vague"; prop: "erty" };

// Type '{ type: "greet"; }' is not assignable to type 'A'.
//   Type '{ type: "greet"; }' is missing the following properties from type '{ type: string; some: "vague"; prop: "erty"; }': some, prop(2322)
const a: A = { type: "greet" };
```
An interesting observation that can be made here is that is we change the `type` property of the last member to a literal type (so all types at the `type` property become literal strings) then TS can improve its error messaging here ([TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8AoKUA+UDeVSQFxQCIBzAJwgmAIG4oBbCAZwYEMiJ8HgSBLAOyKgBfZGkzZw7QixAAJCABt5AewJCR6LDkkBybTQZL6+AgDdWAVwjUoYEkrDGIJUKsFVEiAMZLenKM3w4eDEtY1JySiF3IA)): 
```ts
type A =
  | { type: "greet"; message: string }
  | { type: "sayHello" }
  | { type: ''; some: "vague"; prop: "erty" };

// Type '{ type: "greet"; }' is not assignable to type 'A'.
//   Property 'message' is missing in type '{ type: "greet"; }' but required in type '{ type: "greet"; message: string; }'.(2322)
const a: A = { type: "greet" };
```

Anyway... I've tried to fix the problem in our codebase but I couldn't. It seems that it's somehow related to the fact that `createMachine` has overloads because once I remove the first overload (the model-related one) then it works like intended:
<img width="487" alt="Screenshot 2021-08-20 at 11 04 45" src="https://user-images.githubusercontent.com/9800850/130209384-30633220-eff1-48f0-843f-535c06ea0922.png">
And when I reintroduce the first overload and change it to something non-sensical, that just can't be matched here, like this:
```ts
export function createMachine(config: never): void
```
I just land again with the same problem - the whole `createMachine` gets underlined and a vague error message is reported.

Reply:
I removed the first overload.